File: leetcode/二叉树遍历/0101.py
# 101. 对称二叉树
# https://leetcode-cn.com/problems/symmetric-tree/

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

# [1,2,2,3,4,4,3] - true
# [1,2,2,null,3,null,3] - false

# 1. bfs
# 执行用时： 40 ms , 在所有 Python3 提交中击败了 89.67% 的用户
# 内存消耗： 13.9 MB , 在所有 Python3 提交中击败了 6.06% 的用户
class Solution:
    def isSymmetric(self, root: TreeNode) -> bool:
        if not root:
            return True

        curr_level = [root]

        while True:
            next_level = []

            for node in curr_level:
                if node:
                    next_level.append(node.left)
                    next_level.append(node.right)

            if next_level:
                lst = list(map(lambda x: x.val if x else None, next_level))
                if lst != lst[::-1]:
                    return False
            else:
                break

            curr_level = next_level

        return True

File: leetcode/二叉树遍历/0101.go
// 101. 对称二叉树
// https://leetcode-cn.com/problems/symmetric-tree/
// https://leetcode-cn.com/problems/symmetric-tree/solution/dui-cheng-er-cha-shu-by-leetcode-solution/

// 2. 方法二：迭代
// 执行用时： 4 ms , 在所有 Go 提交中击败了 74.85% 的用户
// 内存消耗： 3.1 MB , 在所有 Go 提交中击败了 28.57% 的用户
func isSymmetric(root *TreeNode) bool {
    u, v := root, root
    q := []*TreeNode{}
    q = append(q, u)
    q = append(q, v)
    for len(q) > 0 {
        u, v = q[0], q[1]
        q = q[2:]
        if u == nil && v == nil {
            continue
        }
        if u == nil || v == nil {
            return false
        }
        if u.Val != v.Val {
            return false
        }
        q = append(q, u.Left)
        q = append(q, v.Right)

        q = append(q, u.Right)
        q = append(q, v.Left)
    }
    return true
}


// 1. 方法一：递归
// 执行用时： 4 ms , 在所有 Go 提交中击败了 74.85% 的用户
// 内存消耗： 2.9 MB , 在所有 Go 提交中击败了 100.00% 的用户
func isSymmetric(root *TreeNode) bool {
    return check(root, root)
}

func check(p, q *TreeNode) bool {
    if p == nil && q == nil {
        return true
    }
    if p == nil || q == nil {
        return false
    }
    return p.Val == q.Val && check(p.Left, q.Right) && check(p.Right, q.Left)
}
